src/shogun/multiclass/KNN.cpp

```diff
@@ -123,33 +121,66 @@ SGMatrix<index_t> KNN::nearest_neighbors()
 	distance->precompute_lhs();
 	distance->precompute_rhs();
 
-	//for each test example
-	for (auto i : SG_PROGRESS(range(n)))
+	switch (m_knn_solver)
 	{
-		COMPUTATION_CONTROLLERS
-		//lhs idx 0..num train examples-1 (i.e., all train examples) and rhs idx i
-		distances_lhs(dists,0,m_train_labels.vlen-1,i);
-
-		//fill in an array with 0..num train examples-1
-		for (int32_t j=0; j<m_train_labels.vlen; j++)
-			train_idxs[j]=j;
-
-		//sort the distance vector between test example i and all train examples
-		Math::qsort_index(dists.vector, train_idxs.vector, m_train_labels.vlen);
-
-#ifdef DEBUG_KNN
-		io::print("\nQuick sort query {}\n", i);
-		for (int32_t j=0; j<m_k; j++)
-			io::print("{} ", train_idxs[j]);
-		io::print("\n");
-#endif
-
-		//fill in the output the indices of the nearest neighbors
-		for (int32_t j=0; j<m_k; j++)
-			NN(j,i) = train_idxs[j];
+	case KNN_BRUTE:
+	{
+		//for each test example
+		for (auto i : SG_PROGRESS(range(n)))
+		{
+			COMPUTATION_CONTROLLERS
+			//lhs idx 0..num train examples-1 (i.e., all train examples) and rhs idx i
+			distances_lhs(dists,0,m_train_labels.vlen-1,i);
+
+			//fill in an array with 0..num train examples-1
+			train_idxs.range_fill(0);
+
+			std::pair<float64_t, index_t> pairt[m_train_labels.vlen]; 
+  
+			// Storing the respective array 
+			// elements in pairs. 
+			for (int j = 0; j < m_train_labels.vlen; j++)  
+			{ 
+				pairt[j].first = dists[j]; 
+				pairt[j].second = train_idxs[j]; 
+			} 
+
+			// Sorting the pair array. 
+			sort(pairt, pairt + m_train_labels.vlen); 
+
+			// Modifying original arrays 
+			for (int j = 0; j < m_train_labels.vlen; j++)  
+			{ 
+				dists[j] = pairt[j].first; 
+				train_idxs[j] = pairt[j].second; 
+			} 
+
+			SG_DEBUG("\nQuick sort query {}", i);
+			SG_DEBUG("{}", train_idxs.to_string());
+			SG_DEBUG("");
+
+			SGVector<index_t> nearest_k_train_idxs(train_idxs.vector, m_k, 0);
+
+			//fill in the output the indices of the nearest neighbors
+			NN.set_column(i, nearest_k_train_idxs);
+		}
+
+		distance->reset_precompute();
+		break;
 	}
+	case KNN_KDTREE:
+	{
+		auto lhs = distance->get_lhs();
+		auto kd_tree = std::make_shared<KDTree>(m_leaf_size);
+		kd_tree->build_tree(lhs->as<DenseFeatures<float64_t>>());
```

Comment:
I have a stupid question. If you build the tree in here, as opposed to say in the train method, what do we gain? KD-trees make it fast to *look up* nearest neighbors once precomputed....

Reply:
No, we don't gain anything at all by doing this. I just did this to prevent changing a lot of existing code.

Currently the KD-tree is built in two functions:
i) `nearest_neighbours()` 
ii) when `classify_objects()` (defined in the `KDTreeKNNSolver` class) is called from the `apply_multiclass()` function.

I'm assuming `nearest_neighbours()` will be called before `apply_multiclass()` is called. Hence it won't be possible to use that instance of the KD-tree.
One solution is to build the KD-tree in the `train()` function after checking if the `m_knn_solver` is of type `KNN_KDTree`. However, we'll need to think of some way to store it in some sort of a member variable and then send it to the `classify_objects()` function defined in the `KDTreeKNNSolver` class.

Reply:
I think this calls for some kind of re-design of the KNN classes.

Definitely you want to build the tree in the train method, and then subsequently you use it in the apply method.